File: 卡码/代码随想录/C++/动态规划/dp/02-背包问题/src/01-01背包/LeetCode_1049.cpp
//1049. 最后一块石头的重量 II
#include <iostream>
#include <vector>
using namespace std;

class Solution {
public:
	int lastStoneWeightII(vector<int>& stones) {
		int sum=0;
		for(int stone:stones) sum+=stone;
		int target=sum/2;
		int dp[1501]={0};
		for(int i=0;i<stones.size();i++){
			for(int j=target;j>=stones[i];j--){
				dp[j]=max(dp[j],dp[j-stones[i]]+stones[i]);
			}
		}
		return sum-2*dp[target];
	}
};

int main(){
  system("pause");
  return 0;
}

File: 卡码/代码随想录/Java/codeThoughts/src/main/java/com/kurumi/leetCode/codeThoughts/动态规划/_02_背包问题/_01_01背包/LeetCode_1049.java
package com.kurumi.leetCode.codeThoughts.动态规划._02_背包问题._01_01背包;

/**
 * @Author kurumi
 * @Date 2026-07-03 上午11:52:20
 * @Description 1049. 最后一块石头的重量 II
 */
public class LeetCode_1049 {
    static class Solution {
        public int lastStoneWeightII(int[] stones) {
            int sum = 0;
            for (int stone : stones)
                sum += stone;
            int target = sum / 2;
            int[] dp = new int[1501];
            for (int i = 0; i < stones.length; i++) {
                for (int j = target; j >= stones[i]; j--) {
                    dp[j] = Math.max(dp[j], dp[j - stones[i]] + stones[i]);
                }
            }
            return sum - 2 * dp[target];
        }
    }
}
